src/modeling_library/switch/update.jl

```diff
@@ -0,0 +1,115 @@
+mutable struct SwitchUpdateState{T}
+    weight::Float64
+    score::Float64
+    noise::Float64
+    prev_trace::Trace
+    trace::Trace
+    index::Int
+    discard::ChoiceMap
+    updated_retdiff::Diff
+    SwitchUpdateState{T}(weight::Float64, score::Float64, noise::Float64, prev_trace::Trace) where T = new{T}(weight, score, noise, prev_trace)
+end
+
+function update_recurse_merge(prev_choices::ChoiceMap, choices::ChoiceMap)
+    prev_choice_submap_iterator = get_submaps_shallow(prev_choices)
+    prev_choice_value_iterator = get_values_shallow(prev_choices)
+    choice_submap_iterator = get_submaps_shallow(choices)
+    choice_value_iterator = get_values_shallow(choices)
+    new_choices = DynamicChoiceMap()
+    for (key, value) in prev_choice_value_iterator
+        key in keys(choice_value_iterator) && continue
+        set_value!(new_choices, key, value)
+    end
+    for (key, node1) in prev_choice_submap_iterator
+        if key in keys(choice_submap_iterator)
+            node2 = get_submap(choices, key)
+            node = update_recurse_merge(node1, node2)
+            set_submap!(new_choices, key, node)
+        else
+            set_submap!(new_choices, key, node1)
+        end
+    end
+    for (key, value) in choice_value_iterator
+        set_value!(new_choices, key, value)
+    end
+    sel, _ = zip(prev_choice_submap_iterator...)
+    comp = complement(select(sel...))
+    for (key, node) in get_submaps_shallow(get_selected(choices, comp))
+        set_submap!(new_choices, key, node)
+    end
+    return new_choices
+end
+
+function update_discard(prev_trace::Trace, choices::ChoiceMap, new_trace::Trace)
+    discard = choicemap()
+    prev_choices = get_choices(prev_trace)
+    for (k, v) in get_submaps_shallow(prev_choices)
+        isempty(get_submap(get_choices(new_trace), k)) && continue
+        isempty(get_submap(choices, k)) && continue
+        set_submap!(discard, k, v)
```

Comment:
I think this requires a recursive call.

Even if the new trace has some choice under this key, and the constraints (`choices`) also has some choice under this key, there could also be choices in the new trace that were copied directly from the previous trace and not overwritten by the constraints.

Reply:
You're absolutely right - this was a mistake in my logic.